src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,323 @@
+
+import backend.resource.TurboLabel;
+import util.Utility;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic which handles toggling/highlighting of labels and simplifies the retrieval of
+ * the resulting status for UI to display.
+ */
+public class LabelPickerState {
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    List<String> repoLabels;
+    OptionalInt currentSuggestionIndex;
+
+    public LabelPickerState(Set<String> initialLabels, List<String> repoLabels) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), repoLabels,
+                OptionalInt.empty());
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<String> repoLabels) {
+        this(initialLabels, addedLabels, removedLabels, matchedLabels, repoLabels, OptionalInt.empty());
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<String> repoLabels, OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.repoLabels = repoLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Determines resulting state based on given initial state and user input
+     *
+     * @param initialState the state of the label picker before the user has typed anything
+     * @param userInput 
+     * @return new state that corresponds with the user input
+     */
+    public static LabelPickerState determineState(LabelPickerState initialState,
+                                                  String userInput) {
+        LabelPickerState state = initialState;
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            state = state.toggleLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            state = state.updateMatchedLabels(keywordInProgess.get());
+        }
+
+        return state;
+    }
+
+    /**
+     * Gives a new state with the label that contains keyword toggled.
+     *
+     * This will simply return the same state if there are more than 1
+     * labels that contain the keyword
+     *
+     * keyword is case-insensitive
+     *
+     * @param keyword
+     * @return
+     */
+    public LabelPickerState toggleLabel(String keyword) {
```

Comment:
This is required to handle the clicking of labels (called by LabelPickerDialog).